### ISSUE TITLE
[Enhancement]: Support different docker context

### ISSUE THREAD

**Issue by diegolovison** [testcontainers/testcontainers-java#5837]
### Module

Core

### Proposal

```
% docker context ls
NAME                  DESCRIPTION                               DOCKER ENDPOINT                                  KUBERNETES ENDPOINT                                           ORCHESTRATOR
default               Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                      https://api.xpto.net:6443 (dlovison)   swarm
desktop-linux                                                   unix:///Users/dlovison/.docker/run/docker.sock                                                                 
my-remote-engine2 *                                             ssh://vagrant@192.168.56.6                                       
```

When I execute my test
```
16:25:20,932 ERROR (testng-ExampleJdg4931) [org.testcontainers.dockerclient.DockerClientProviderStrategy] Could not find a valid Docker environment. Please check configuration. Attempted configurations were:
16:25:20,932 ERROR (testng-ExampleJdg4931) [org.testcontainers.dockerclient.DockerClientProviderStrategy]     UnixSocketClientProviderStrategy: failed with exception InvalidConfigurationException (Could not find unix domain socket). Root cause NoSuchFileException (/var/run/docker.sock)
16:25:20,933 ERROR (testng-ExampleJdg4931) [org.testcontainers.dockerclient.DockerClientProviderStrategy] As no valid configuration was found, execution cannot continue

```

**Comment by diegolovison** [comment]
Require https://github.com/docker-java/docker-java/issues/1946

**Comment by eddumelendez** [comment]
@diegolovison I think you are using Colima, right? Check the [docs](https://www.testcontainers.org/supported_docker_environment/#using-colima) on how to set the env vars.

**Comment by t1** [comment]
It would be very nice, if Colima would be detected automatically, just like Docker-Desktop is. I assume this is not a political thing 😘

**Comment by skagedal** [comment]
@t1 I think this assumption is correct. My efforts so far about implementing support for Docker contexts have been well received by the maintainers. I am working on a pull request for docker-java [here](https://github.com/docker-java/docker-java/pull/2036). Hope to be able to finish it up soon. I am also blogging a bit about my progress, [starting here](https://blog.skagedal.tech/2023/01/30/test-containers-and-colima.html).

**Comment by chriskilding** [comment]
Is there any update on this? The relevant PR in docker-java was merged a couple of weeks ago and I think they've done a release 3.3.0 with that change included.

Do we just need a new release of testcontainers-java to pick this change up?

**Comment by skagedal** [comment]
Oh cool, I missed that – indeed [3.3.0](https://github.com/docker-java/docker-java/releases/tag/3.3.0) seems to include it. And the dependency in testcontainers-java [was also bumped](https://github.com/testcontainers/testcontainers-java/commit/2c4caadc3d88bea2e75f9b83cbfe122995ca9bb5). So yeah, I guess that's what we need!

**Comment by chriskilding** [comment]
@eddumelendez would you be able to publish a new release? (Alternatively, when does the Testcontainers project have a release scheduled next?)

**Comment by piotrwielgolaski-tomtom** [comment]
I build custom release locally, but it does not work out of box, so anyway some adjustment need to be done to make it work before release

**Comment by skagedal** [comment]
@piotrwielgolaski-tomtom Could you provide some details on what you tried and in what you saw?

**Comment by piotrwielgolaski-tomtom** [comment]
@skagedal my scenario was simple, use docker context to connect with remote docker instance. 
when I rebuild latest code I wanted to give a try, but it don't work. 
it seems that test containers have concept of DockerClientProviderStrategy, but any of existing one will take into account context set by docker, maybe EnvironmentAndSystemPropertyClientProviderStrategy requires same extension or extra strategy need to be added

**Comment by piotrwielgolaski-tomtom** [comment]
I take another swing on this and simple change in EnvironmentAndSystemPropertyClientProviderStrategy https://github.com/piotrwielgolaski-tomtom/testcontainers-java/blob/de1324ed2800eff4da326d0c23d281399d006bc0/core/src/main/java/org/testcontainers/dockerclient/EnvironmentAndSystemPropertyClientProviderStrategy.java#L51

```
            case "autoIgnoringUserProperties":
                dockerClientConfig = configBuilder.build();
                applicable = dockerClientConfig.getDockerHost() != null;
                break;
```
make it work a bit more, but you need to opt in for `dockerconfig.source=autoIgnoringUserProperties`

@skagedal I noticed that you focused on docker host, but all TLS is missing, so when I use remote connection over tcp with tls it will not work. It would be lovely if docker-java also support resolution of getSSLConfig based on context

**Comment by skagedal** [comment]
@piotrwielgolaski-tomtom Right, there is a `SkipTLSVerify` boolean property in the context that also should have been taken into account. I missed/forgot that. When you say "all TLS", do you mean other things as well?

**Comment by skagedal** [comment]
Oh, there is a `TLSMaterial` property... https://docs.docker.com/engine/context/working-with-contexts/ 

I never found any complete documentation on this format, does anyone know what that might be?

**Comment by piotrwielgolaski-tomtom** [comment]
@skagedal The best that I can find https://github.com/docker/cli/blob/a0756c3c2cacebf5e5dc6454cc280c3ddf675176/docs/reference/commandline/context_inspect.md?plain=1#L42
I thin that this path if exist should be used for SSL config, of course if skip is not true.

**Comment by skagedal** [comment]
And what would it expect to find in that path?

**Comment by piotrwielgolaski-tomtom** [comment]
> And what would it expect to find in that path?

I would expect that whatever docker expects to configure SSL https://github.com/skagedal/docker-java/blob/88cea47d18cce4339a202ae7ad66fe68ee00d533/docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerClientConfig.java#LL450C58-L450C72
It should be directory with certificate

**Comment by piotrwielgolaski-tomtom** [comment]
It does not work out of box, I wrote details above. 
* to make it work with docker host and TLS it requires https://github.com/docker-java/docker-java/pull/2105
* ignoring this it still need extra handling in test containers - snippet above

**Comment by glasser** [comment]
Ah cool — is there a PR open for that last bit? Is that something I can work around as a user without forking testcontainers?

**Comment by glasser** [comment]
While I don't know if this supports all context use cases, #6927 is sufficient to support Colima without any configuration.

**Comment by g0t4** [comment]
FYI in the latest build of `Docker Desktop for macOS` it no longer automatically sets/updates `/var/run/docker.sock`
- To re-enable use `Settings` > `Advanced` > `Allow the default Docker socket to be used (requires password)`. 
- If `/var/run/docker.sock` isn't valid then `testcontainers` fails with:

```
Could not find a valid Docker environment. Please check configuration. Attempted configurations were: ...
(Could not find unix domain socket). Root cause NoSuchFileException (/var/run/docker.sock)
```

If `testcontainers` supported contexts, I could use `docker context use desktop-linux` which points to a user specific socket: `unix:///Users/$USER/.docker/run/docker.sock`

**Comment by poklakni** [comment]
what is the current status?

**Comment by OscarPalafox** [comment]
In https://github.com/testcontainers/testcontainers-java/pull/5039, @kiview and @bsideup both mention it is possible to do it with custom Docker configurations. Would it be possible to please get some guidance on how to achieve this?

**Comment by mikiTesf** [comment]
[This](https://github.com/testcontainers/testcontainers-java/issues/5034#issuecomment-1036437136) could be related

**Comment by fransf-anva** [comment]
It would seem the root of the issue is that there is some duplication between testcontainers and docker-java, as confirmed by [this TODO](https://github.com/testcontainers/testcontainers-java/blob/main/core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java#L153) in `DockerClientProviderStrategy`:
```
    /**
     * TODO we should consider moving this to docker-java at some point
     */
    @UnstableAPI
    protected boolean test() {
```

If testcontainers were to just delegate the entire process for checking whether Docker is running and whether it can connect to Docker, this should pretty much work out of the box, right? 

And that seems like the right thing to do: if we're delegating all the details of connecting to a Docker daemon to Docker-java, then testcontainers shouldn't take it upon itself to try and connect to a TCP or Unix socket just to determine whether a Docker daemon is accessible: that, too, should be handled by Docker-java. And that would then work regardless of how the connection to Docker is actually made, including whether a default context is found and used.

If that's the case, this seems like a fairly simple case of re-implementing the `test()` method on [DockerClientProviderStrategy](https://github.com/testcontainers/testcontainers-java/blob/main/core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java). A naive approach might be to just call `getDockerClient()` and if it doesn't throw an exception, return `true`, otherwise return `false`.

I'd be happy to have a go at a PR for that. Any pointers or things I should be aware of?

I saw a note on a different issue about following what the NodeJS testcontainers had done. Is there a summary of what that is?


I think it's a reasonable ask that if you can type `docker ps` on your command line and see a Docker container list, you should be able to run testcontainers from a Java project build on that same command line without having to set additional environment variables.
